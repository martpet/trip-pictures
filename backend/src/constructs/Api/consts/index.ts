export * from './authorizationHeader';
export * from './lambdaEdgeViewerEvent';
