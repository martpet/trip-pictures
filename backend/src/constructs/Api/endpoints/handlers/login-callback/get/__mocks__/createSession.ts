export const createSession = vi
  .fn()
  .mockName('createSession')
  .mockResolvedValue('dummySessionCookie');
