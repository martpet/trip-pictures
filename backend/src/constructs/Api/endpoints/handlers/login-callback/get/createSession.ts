import {
  cookie,
  cookieName,
  crypto,
  DynamoDBClient,
  DynamoDBDocumentClient,
  millis,
  OauthTokens,
  PutCommand,
  refreshTokenValidityInDays,
  SessionsTableItem,
  sessionsTableOptions,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocumentClient.from(ddbClient);

type CreateSessionProps = {
  tokens: OauthTokens;
};

export const createSession = async ({ tokens }: CreateSessionProps) => {
  const { idTokenPayload, refreshToken, idToken } = tokens;
  const { envName } = globalLambdaProps;
  const sessionId = crypto.randomUUID();
  const created = Date.now();
  const refreshTokenExpires = (created + millis.days(refreshTokenValidityInDays)) / 1000;

  const putCommand = new PutCommand({
    TableName: sessionsTableOptions.tableName,
    Item: {
      id: sessionId,
      userId: idTokenPayload.sub,
      created,
      refreshToken,
      refreshTokenExpires,
      idToken,
    } satisfies SessionsTableItem,
  });

  await ddbDocClient.send(putCommand);

  const sessionCookie = cookie.serialize(cookieName('sessionId'), sessionId, {
    secure: true,
    httpOnly: true,
    sameSite: envName !== 'personal' || 'none',
    expires: new Date(refreshTokenExpires * 1000),
  });

  return sessionCookie;
};
