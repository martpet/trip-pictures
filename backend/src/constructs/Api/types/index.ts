export * from './Cookies';
export * from './EnvVars';
export * from './LambdaEdgeOriginRequestHandler';
export * from './LambdaEdgeResponse';
export * from './OauthCookieProps';
export * from './OauthTokens';
export * from './RouteHeaders';
