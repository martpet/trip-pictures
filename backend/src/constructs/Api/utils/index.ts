export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
