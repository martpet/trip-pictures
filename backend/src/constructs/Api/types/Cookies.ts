import { ConditionalKeys } from 'type-fest';

import { apiOptions } from '~/consts';
import { ApiRouteOptions } from '~/types';

export type PathWithCookies = ConditionalKeys<
  typeof apiOptions,
  Pick<ApiRouteOptions, 'cookies'>
>;

export type Cookies<T extends PathWithCookies> = Partial<
  Record<(typeof apiOptions)[T]['cookies'][number], string>
>;

export type AllCookies = Cookies<PathWithCookies>;
