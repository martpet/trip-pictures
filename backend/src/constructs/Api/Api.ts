import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getCdkEnv } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getCdkEnv();
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'Distro-Log-Bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'Distro', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      distribution,
      origin,
    });

    new ARecord(this, 'A-Record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
