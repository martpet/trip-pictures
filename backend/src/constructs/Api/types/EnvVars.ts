import { SetRequired } from 'type-fest';

import { apiOptions } from '~/consts';
import { ApiMethodOptions, DeepValues } from '~/types';

type A = typeof apiOptions;

export type EnvVars<
  R extends keyof A,
  M extends keyof A[R]['methods'],
  O = A[R]['methods'][M]
> = O extends SetRequired<ApiMethodOptions, 'envVars'>
  ? Partial<Record<O['envVars'][number], string>>
  : {};

export type AllEnvVars = Record<DeepValues<A, 'envVars'>[number], string>;
