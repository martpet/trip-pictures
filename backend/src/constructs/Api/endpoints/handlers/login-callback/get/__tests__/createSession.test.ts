import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import cookie from 'cookie';

import { OauthTokens } from '~/constructs/Api/types';
import { itCalls, itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { createSession } from '../createSession';

const ddbMock = mockClient(DynamoDBDocumentClient);

vi.mock('cookie');
vi.mock('crypto');

const args = [
  {
    tokens: {
      accessToken: 'dummyAccessToken',
      refreshToken: 'dummyRefreshToken',
      idToken: 'dummyIdToken',
      idTokenPayload: { sub: 'dummySub ' },
    } as OauthTokens,
  },
] as Parameters<typeof createSession>;

beforeEach(() => {
  ddbMock.reset();
});

describe('createSession', () => {
  itSendsAwsCommand(PutCommand, ddbMock, createSession, args);
  itCalls(cookie.serialize, createSession, args);
  itResolves(createSession, args);

  describe('when "globalLambdaProps.envName" is "personal"', () => {
    const initialEnvName = globalLambdaProps.envName;
    beforeAll(() => {
      globalLambdaProps.envName = 'personal';
    });
    afterAll(() => {
      globalLambdaProps.envName = initialEnvName;
    });
    itCalls(cookie.serialize, createSession, args);
  });
});
