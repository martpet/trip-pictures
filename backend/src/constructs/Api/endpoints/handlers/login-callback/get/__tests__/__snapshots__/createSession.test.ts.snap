// Vitest Snapshot v1

exports[`createSession > calls "serialize" with correct args 1`] = `
[
  [
    "sessionId",
    "dummyUUID",
    {
      "expires": 2023-01-01T00:00:00.000Z,
      "httpOnly": true,
      "sameSite": true,
      "secure": true,
    },
  ],
]
`;

exports[`createSession > resolves with a correct value 1`] = `"dummyCookie"`;

exports[`createSession > sends "PutCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "Item": {
    "created": 1640995200000,
    "id": "dummyUUID",
    "idToken": "dummyIdToken",
    "refreshToken": "dummyRefreshToken",
    "refreshTokenExpires": 1672531200,
    "userId": "dummySub ",
  },
  "TableName": "TP-sessions",
}
`;

exports[`createSession > when "globalLambdaProps.envName" is "personal" > calls "serialize" with correct args 1`] = `
[
  [
    "sessionId",
    "dummyUUID",
    {
      "expires": 2023-01-01T00:00:00.000Z,
      "httpOnly": true,
      "sameSite": "none",
      "secure": true,
    },
  ],
]
`;
